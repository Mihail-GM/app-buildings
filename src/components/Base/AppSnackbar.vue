<template>
	<div class="text-center">
		<v-snackbar
			color="deep-purple accent-4"
			class="mb-15"
			elevation="12"
			v-model="showSnackbar"
			:multi-line="multiLine"
			:timeout="snackbarMessageData.duration"
			id="snackbar-message"
		>
			{{ snackbarMessageData.message }}

			<template v-slot:action="{ attrs }">
				<v-btn
					id="snackbar-close-button"
					text
					v-bind="attrs"
					@click="closeSnackbar()"
				>
					Close
				</v-btn>
			</template>
		</v-snackbar>
	</div>
</template>

<script>
	import { mapGetters } from "vuex";

	export default {
		name: "AppSnackbar",

		data: () => ({
			multiLine: false,
			showSnackbar: false,
		}),

		methods: {
			closeSnackbar() {
				this.showSnackbar = false;
			},
		},

		computed: {
			...mapGetters(
				{
					snackbarMessageData: 'snackbar/getSnackbarMessage'
				}
			)
		},

		watch: {
			snackbarMessageData() {
				this.showSnackbar = true;
			},
		},
	};

</script>

<style lang="scss" scoped>

</style>