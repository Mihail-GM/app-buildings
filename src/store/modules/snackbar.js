const state = {
	snackbarMessageData: {
		message: '',
		duration: 3000,
		mode: 'success'
	},
};

const getters = {
	getSnackbarMessage(state) {
		return state.snackbarMessageData;
	},
};

const mutations = {
	setSnackbarMessage(state, {message, duration = 4000, mode = 'success', buttonText = 'Close'}) {
		state.snackbarMessageData = {message, duration, mode, buttonText};
	},
}

const actions = {
	showSnackbarMessage(vuexContext, messageData) {
		vuexContext.commit('setSnackbarMessage', messageData);
	},
}

export default {
	namespaced: true,
	state,
	getters,
	mutations,
	actions
}