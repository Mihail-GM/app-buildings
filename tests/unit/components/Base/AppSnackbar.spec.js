import Vuetify from 'vuetify'
import Vuex from "vuex";

import {createLocalVue, mount} from '@vue/test-utils'
import AppSnackbar from "@/components/Base/AppSnackbar";
import {merge} from "lodash";
import snackbar from "@/store/modules/snackbar";
import flushPromises from "flush-promises";

describe('AppSnackbar.vue', () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);
	let vuetify

	beforeEach(() => {
		vuetify = new Vuetify()
	})

	function createStore(overrides) {
		let snackbarMessageData = {
			message: 'Test',
			duration: 3000,
			mode: 'success'
		};

		const defaultStoreConfig = {
			modules: {
				snackbar: {
					namespaced: true,
					state: {
						snackbarMessageData
					},
					getters: snackbar.getters,
					mutations: snackbar.mutations,
					actions: snackbar.actions,
				}
			}
		};

		return new Vuex.Store(merge(defaultStoreConfig, overrides));
	}

	const createWrapper = (overrides) => {
		const defaultMountingOptions = {
			localVue,
			vuetify,
			store: createStore(),
		};

		const dataOptions = {
			data() {
				return {
					showSnackbar: true
				}
			}
		};

		return mount(AppSnackbar, merge(defaultMountingOptions, dataOptions, overrides));
	}

	it('should close dialog when close button is clicked', () => {
		const wrapper = createWrapper();
		expect(wrapper.vm.showSnackbar).toBe(true);

		const button = wrapper.find('#snackbar-close-button');
		button.trigger('click');

		expect(wrapper.vm.showSnackbar).toBe(false);
	})

	it('change message when new message is set to store', async () => {
		let store = createStore();
		const wrapper = createWrapper(store);

		await wrapper.vm.$store.dispatch('snackbar/showSnackbarMessage', {
			message: 'Changed.',
			duration: 4000,
			mode: 'success'
		});
		await flushPromises;

		const element = wrapper.find('.v-snack__content');
		expect(element.text()).toBe('Changed.');
	})

	it('should match snapshot', () => {
		const wrapper = createWrapper()

		expect(wrapper.html()).toMatchSnapshot()
	})
})
