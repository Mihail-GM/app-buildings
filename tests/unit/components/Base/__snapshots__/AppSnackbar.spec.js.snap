// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppSnackbar.vue should match snapshot 1`] = `
<div class="text-center">
  <div class="v-snack mb-15 v-snack--active v-snack--bottom v-snack--has-background" id="snackbar-message">
    <transition-stub name="v-snack-transition">
      <div class="v-snack__wrapper v-sheet theme--dark elevation-12 deep-purple accent-4">
        <div role="status" aria-live="polite" class="v-snack__content">
          Test

        </div>
        <div class="v-snack__action "><button type="button" class="v-btn v-btn--text theme--dark v-size--default v-snack__btn" id="snackbar-close-button"><span class="v-btn__content">
				Close
			</span></button></div>
      </div>
    </transition-stub>
  </div>
</div>
`;
